cuda/solver/common_trs_kernels.cuh

```diff
@@ -0,0 +1,240 @@
+                solve_struct->policy, &solve_struct->factor_work_size);
+
+            // allocate workspace
+            if (solve_struct->factor_work_vec != nullptr) {
+                exec->free(solve_struct->factor_work_vec);
+            }
+            solve_struct->factor_work_vec =
+                exec->alloc<void *>(solve_struct->factor_work_size);
+
+            cusparse::csrsm2_analysis(
+                handle, solve_struct->algorithm,
+                CUSPARSE_OPERATION_NON_TRANSPOSE, CUSPARSE_OPERATION_TRANSPOSE,
+                matrix->get_size()[0], num_rhs,
+                matrix->get_num_stored_elements(), &one,
+                solve_struct->factor_descr, matrix->get_const_values(),
+                matrix->get_const_row_ptrs(), matrix->get_const_col_idxs(),
+                nullptr, num_rhs, solve_struct->solve_info,
+                solve_struct->policy, solve_struct->factor_work_vec);
+        }
+
+
+#elif (defined(CUDA_VERSION) && (CUDA_VERSION < 9020))
+
+
+        {
+            cusparse::pointer_mode_guard pm_guard(handle);
+            cusparse::csrsm_analysis(
+                handle, CUSPARSE_OPERATION_NON_TRANSPOSE, matrix->get_size()[0],
+                matrix->get_num_stored_elements(), solve_struct->factor_descr,
+                matrix->get_const_values(), matrix->get_const_row_ptrs(),
+                matrix->get_const_col_idxs(), solve_struct->solve_info);
+        }
+
+
+#endif
+
+
+    } else {
+        GKO_NOT_IMPLEMENTED;
+    }
+}
+
+
+template <typename ValueType, typename IndexType>
+void solve_kernel(std::shared_ptr<const CudaExecutor> exec,
+                  const matrix::Csr<ValueType, IndexType> *matrix,
+                  const solver::SolveStruct *solve_struct,
+                  matrix::Dense<ValueType> *trans_b,
+                  matrix::Dense<ValueType> *trans_x,
+                  const matrix::Dense<ValueType> *b,
+                  matrix::Dense<ValueType> *x)
+{
+    using vec = matrix::Dense<ValueType>;
+    if (cusparse::is_supported<ValueType, IndexType>::value) {
+        ValueType one = 1.0;
+        auto handle = exec->get_cusparse_handle();
+
+
+#if (defined(CUDA_VERSION) && (CUDA_VERSION >= 9020))
+
+
+        x->copy_from(gko::lend(b));
+        {
+            cusparse::pointer_mode_guard pm_guard(handle);
+            cusparse::csrsm2_solve(
+                handle, solve_struct->algorithm,
+                CUSPARSE_OPERATION_NON_TRANSPOSE, CUSPARSE_OPERATION_TRANSPOSE,
+                matrix->get_size()[0], b->get_stride(),
+                matrix->get_num_stored_elements(), &one,
+                solve_struct->factor_descr, matrix->get_const_values(),
+                matrix->get_const_row_ptrs(), matrix->get_const_col_idxs(),
+                x->get_values(), b->get_stride(), solve_struct->solve_info,
+                solve_struct->policy, solve_struct->factor_work_vec);
+        }
+
+#elif (defined(CUDA_VERSION) && (CUDA_VERSION < 9020))
```

Comment:
nit: since you always put 2 newlines before and after every preprocessor `if`, over this `elif`, there is only one.